inc/dashboard/namespace.php

```diff
@@ -0,0 +1,436 @@
+ *
+ * @param array $buckets The ElasticSearch query data.
+ *
+ * @return array An array of aggregated data.
+ */
+function get_aggregate_data( array $buckets ) : array {
+	$data = [];
+
+	foreach ( $buckets as $block ) {
+		$block_id = $block['key'];
+		$block_data = wp_list_pluck( $block['events']['buckets'], 'doc_count', 'key' );
+		$views = $block_data['experienceView'] ?? 0;
+		$conversions = $block_data['conversion'] ?? 0;
+		$data[ $block_id ]['views'] = $views;
+		$data[ $block_id ]['conversions'] = $conversions;
+		$data[ $block_id ]['avg_conversion_rate'] = calculate_average_conversion_rate( [
+			'views' => $views,
+			'conversions' => $conversions,
+		] );
+	}
+
+	return $data;
+}
+
+/**
+ * Sort data by conversion rate.
+ *
+ * @param array $list The array of analytics data by block id.
+ * @param string $order The order to sort by. Accepted values are 'asc' or 'desc'.
+ *
+ * @return array The sorted array.
+ */
+function sort_by_conversion_rate( array $list, string $order = 'desc' ) : array {
+	// If an invalid value was passed to $order, default to 'desc'.
+	if ( ! in_array( $order, [ 'asc', 'desc' ], true ) ) {
+		$order = 'desc';
+	}
+
+	$avg_conversion_rate = array_column( $list, 'avg_conversion_rate' );
+	$sort_order = 'SORT_' . strtoupper( $order );
+
+	array_multisort( $avg_conversion_rate, $sort_order, $list );
+
+	return $list;
+}
+
+/**
+ * Get data from a single block.
+ *
+ * @uses get_views_List()
+ *
+ * @param string $block_id Get a single block's analytics data pulled from the ES query.
+ *
+ * @return array The block's analytics data.
+ */
+function get_block_data( string $block_id ) : array {
+	$data = get_views_list();
+	return $data[ $block_id ];
+}
+
+/**
+ * Get analytics views list data.
+ *
+ * @todo Integrate date range searches.
+ *
+ * @param string $order How to order the data. Accepted values are 'asc' and 'desc'. Defaults to 'desc'.
+ * @param int $days How many days worth of analytics to fetch. Defaults to 7.
+ */
+function get_views_list( string $order = 'desc', int $days = 7 ) : array {
+	$query = [
+		'query' => [
+			'bool' => [
+				'filter' => [
+					[
+						// Query from the current site.
+						'term' => [
+							'attributes.blogId.keyword' => get_current_blog_id(),
+						],
+					],
+					[
+						// We're just interested in views.
+						'terms' => [
+							'event_type.keyword' => [
+								'experienceView',
+								'conversion',
+							],
+						],
+					],
+				],
+			],
+		],
+		'aggs' => [
+			'blocks' => [
+				'terms' => [
+					// Get the block data. This will give us the key for the block, which is stored as the post slug.
+					'field' => 'attributes.clientId.keyword',
+					'size' => 10000, // Use arbitrary large size that is more than we're likely to need.
+					'order' => [
+						'_count' => $order,
+					],
+				],
+				'aggs' => [
+					'events' => [
+						'terms' => [
+							'field' => 'event_type.keyword',
+						],
+					],
```

Comment:
```suggestion
					],
					'aggs' => [
						'uniques' => [
							'cardinality' => 'endpoint.Id.keyword',
						],
					],
```

Reply:
This throws an error after I merged it in.

```Analytics: elasticsearch query failed: http://elasticsearch:9200/analytics*/_search {"query":{"bool":{"filter":[{"term":{"attributes.blogId.keyword":1}},{"terms":{"event_type.keyword":["experienceView","conversion"]}}]}},"aggs":{"blocks":{"terms":{"field":"attributes.clientId.keyword","size":10000,"order":{"_count":"asc"}},"aggs":{"events":{"terms":{"field":"event_type.keyword"}},"aggs":{"uniques":{"cardinality":"endpoint.Id.keyword"}}}}},"size":0} {"error":{"root_cause":[{"type":"named_object_not_found_exception","reason":"[1:318] unable to parse BaseAggregationBuilder with name [uniques]: parser not found"}],"type":"named_object_not_found_exception","reason":"[1:318] unable to parse BaseAggregationBuilder with name [uniques]: parser not found"},"status":400}```

Reply:
Oh whoops, that sub `aggs` should be inside the terms aggregation for events, at the same level as the `field` property. My fault

Reply:
Hmm...that's giving me a different error.
```
Analytics: elasticsearch query failed: http://elasticsearch:9200/analytics*/_search {"query":{"bool":{"filter":[{"term":{"attributes.blogId.keyword":1}},{"terms":{"event_type.keyword":["experienceView","conversion"]}}]}},"aggs":{"blocks":{"terms":{"field":"attributes.clientId.keyword","size":10000,"order":{"_count":"asc"}},"aggs":{"events":{"terms":{"field":"event_type.keyword","aggs":{"uniques":{"cardinality":"endpoint.Id.keyword"}}}}}}},"size":0} {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"[terms] unknown field [aggs], parser not found"}],"type":"illegal_argument_exception","reason":"[terms] unknown field [aggs], parser not found"},"status":400}
```